File: worker.cpp
// worker.cpp  (robust Linux-only version)
// Purpose:
//   - Behaves like a daemonized worker (no stdout prints)
//   - Sends job words over UDP multicast (one word every 250 ms)
//   - Sends lifecycle logs via UDP "syslog" messages to localhost
//
// Key design notes:
//   * We DO NOT actually daemonize (per assignment). We just avoid stdout.
//   * We DO log to a UDP "syslog" port (default 30001) to simulate daemon logs.
//   * Multicast sender socket is created via the provided multicast.h helper.
//   * The work-queue address is parsed but not used in Step 3 (placeholder for Step 4).

#include <arpa/inet.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <unistd.h>

#include <csignal>
#include <chrono>
#include <cstring>
#include <optional>
#include <sstream>
#include <string>
#include <thread>

#include "multicast.h"  // mulitcastSenderSocket(), multicastReceiverSocket(...)

// ======= Configuration defaults =======
static constexpr const char* kMulticastGroup = "239.0.0.1"; // IPv4 multicast range: 224.0.0.0 - 239.255.255.255
static constexpr uint16_t    kDefaultOutPort = 30000;       // where words are sent
static constexpr uint16_t    kDefaultLogPort = 30001;       // where "syslog" logs are sent (UDP, localhost)
static constexpr auto        kWordInterval   = std::chrono::milliseconds(250);

// ======= Global for graceful Ctrl+C =======
static volatile std::sig_atomic_t g_stop = 0;
static void handle_sigint(int) { g_stop = 1; }

// ======= Small utilities (no stdout; errors go to syslog) =======

// Quick port sanity
static inline bool is_valid_port(uint16_t p) { return p != 0; }

// Wrap a UDP "syslog-like" sender. Keeps one socket open for all logs to avoid per-call socket churn.
class SyslogSender {
public:
    explicit SyslogSender(uint16_t port) : ok_(false) {
        // Create UDP socket
        s_ = ::socket(AF_INET, SOCK_DGRAM, 0);
        if (s_ < 0) return;

        std::memset(&dst_, 0, sizeof(dst_));
        dst_.sin_family = AF_INET;
        dst_.sin_port   = htons(port);
        if (::inet_pton(AF_INET, "127.0.0.1", &dst_.sin_addr) != 1) {
            ::close(s_);
            s_ = -1;
            return;
        }
        ok_ = true;
    }

    ~SyslogSender() {
        if (s_ >= 0) ::close(s_);
    }

    bool ok() const { return ok_; }

    // RFC3164-ish minimal payload: <14>app[pid]: message
    void log(const std::string& app, const std::string& message) const {
        if (!ok_) return;
        std::ostringstream oss;
        oss << "<14>" << app << "[" << ::getpid() << "]: " << message;
        const std::string payload = oss.str();
        (void)::sendto(s_, payload.data(), (int)payload.size(), 0,
                       reinterpret_cast<const sockaddr*>(&dst_), sizeof(dst_));
    }

private:
    int         s_  = -1;
    bool        ok_ = false;
    sockaddr_in dst_{};
};

// Build an IPv4 sockaddr_in from (addr string, port). Returns std::nullopt on error.
static std::optional<sockaddr_in> make_ipv4_dest(const char* addr, uint16_t port) {
    sockaddr_in out{};
    out.sin_family = AF_INET;
    out.sin_port   = htons(port);
    if (::inet_pton(AF_INET, addr, &out.sin_addr) != 1) {
        return std::nullopt;
    }
    return out;
}

// Check if an IPv4 address is in the multicast block 224.0.0.0/4 (first octet 224..239)
static bool is_multicast_addr(const in_addr& a) {
    const uint8_t first_octet = static_cast<uint8_t>(a.s_addr & 0xFF);
    return first_octet >= 224 && first_octet <= 239;
}

// Trim helper (for stdin-based jobs that might end with newlines)
static inline void rtrim_inplace(std::string& s) {
    while (!s.empty() && (s.back() == '\n' || s.back() == '\r' || s.back() == ' ' || s.back() == '\t'))
        s.pop_back();
}

// ======= Main =======
int main(int argc, char** argv) {
    // Handle Ctrl+C gracefully
    std::signal(SIGINT, handle_sigint);

    // ---- Parse CLI ----
    // Args:
    //   1) workQueueHost:port  (placeholder for Step 4; not used in Step 3)
    //   2) outputPort          (UDP multicast port for words)
    //   3) syslogPort          (UDP port where we send logs to 127.0.0.1)
    //   4+) job text           (if omitted, read from stdin)
    const std::string workQueue = (argc >= 2) ? argv[1] : "hawk.cs.umanitoba.ca:50001";

    const uint16_t outPort = (argc >= 3) ? static_cast<uint16_t>(std::stoi(argv[2])) : kDefaultOutPort;
    const uint16_t logPort = (argc >= 4) ? static_cast<uint16_t>(std::stoi(argv[3])) : kDefaultLogPort;

    // Initialize "syslog" sender ASAP so we can report errors there.
    SyslogSender logger(logPort);
    // Note: if logger.ok() is false, we silently continue (no stdout per assignment).

    // Port validation (log and exit if invalid)
    if (!is_valid_port(outPort) || !is_valid_port(logPort)) {
        if (logger.ok()) logger.log("worker", "invalid port argument(s)");
        return 1;
    }

    // Build job text from args or stdin
    std::string job;
    if (argc >= 5) {
        for (int i = 4; i < argc; ++i) {
            if (i > 4) job.push_back(' ');
            job += argv[i];
        }
    } else {
        std::ostringstream oss;
        oss << std::cin.rdbuf();
        job = oss.str();
    }
    rtrim_inplace(job);
    if (job.empty()) {
        job = "default demo job text"; // safe fallback for smoke testing
    }

    // Start banner (goes to "syslog")
    if (logger.ok()) {
        std::ostringstream oss;
        oss << "starting with workQueue=" << workQueue
            << " multicast=" << kMulticastGroup << ":" << outPort
            << " syslogPort=" << logPort;
        logger.log("worker", oss.str());
    }

    // ---- Prepare multicast destination ----
    // Validate multicast group is well-formed and actually in the multicast range.
    auto maybeDest = make_ipv4_dest(kMulticastGroup, outPort);
    if (!maybeDest.has_value()) {
        if (logger.ok()) logger.log("worker", "invalid multicast group address (inet_pton failed)");
        return 1;
    }
    sockaddr_in dest = *maybeDest;
    if (!is_multicast_addr(dest.sin_addr)) {
        if (logger.ok()) logger.log("worker", "destination is not a multicast address (must be 224.0.0.0/4)");
        return 1;
    }

    // ---- Create multicast sender socket via helper (TTL etc set in multicast.h) ----
    // NOTE: helper name in your header is "mulitcastSenderSocket" (typo preserved).
    int sendSock = mulitcastSenderSocket();
    if (sendSock < 0) {
        if (logger.ok()) logger.log("worker", "failed to create multicast send socket");
        return 1;
    }

    // Lifecycle logs
    if (logger.ok()) logger.log("worker", "fetching job");
    if (logger.ok()) logger.log("worker", "starting job");

    // ---- Main send loop: one word every 250 ms ----
    // We push '\n' for readability in a plain UDP listener like `nc -u -l <port>`.
    std::istringstream iss(job);
    std::string word;
    while (!g_stop && (iss >> word)) {
        word.push_back('\n');
        const ssize_t n = ::sendto(sendSock, word.data(), (int)word.size(), 0,
                                   reinterpret_cast<const sockaddr*>(&dest), sizeof(dest));
        if (n < 0) {
            if (logger.ok()) logger.log("worker", "sendto failed (continuing)");
            // We keep going; a transient error shouldn't kill the worker.
        }
        std::this_thread::sleep_for(kWordInterval);
    }

    if (logger.ok()) logger.log("worker", "completed job");

    ::close(sendSock);
    return 0;
}

File: worker.py
#!/usr/bin/env python3
"""Simple TCP worker for the messaging queue assignment."""

import argparse
import signal
import socket
import sys
import time
from typing import Optional

# Import the multicast library
import multicast

FETCH_INTERVAL = 0.5  # seconds to wait before asking for more work when idle
WORD_DELAY = 0.25     # seconds to sleep between each word of the job text

# Multicast configuration
MULTICAST_GROUP = "239.0.0.1"

stop_requested = False


def _signal_handler(_sig, _frame) -> None:
    global stop_requested
    stop_requested = True


def recv_line(sock: socket.socket) -> Optional[str]:
    """Read a single newline-delimited line; return None on disconnect."""
    data = bytearray()
    try:
        while True:
            chunk = sock.recv(1)
            if not chunk:
                return None
            if chunk == b"\n":
                return data.decode(errors="ignore")
            data.extend(chunk)
    except OSError:
        return None


def send_line(sock: socket.socket, line: str) -> bool:
    try:
        sock.sendall((line + "\n").encode())
        return True
    except OSError:
        return False


def send_to_syslog(message: str, syslog_port: int) -> None:
    """Send a message to syslog via UDP (broadcast, not multicast)."""
    try:
        with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as sock:
            syslog_msg = f"<14>worker: {message}\n"
            # Send as regular UDP unicast to localhost
            sock.sendto(syslog_msg.encode(), ("127.0.0.1", syslog_port))
    except OSError:
        pass


def process_job(mcast_sock: socket.socket, tcp_sock: socket.socket, job_id: int, 
                text: str, output_port: int, syslog_port: int) -> bool:
    """Process a job by sending each word via multicast."""
    send_to_syslog(f"starting job {job_id}", syslog_port)
    
    for word in text.split():
        # Send word via multicast using the library's socket
        payload = f"{word}\n".encode()
        mcast_sock.sendto(payload, (MULTICAST_GROUP, output_port))
        time.sleep(WORD_DELAY)
        if stop_requested:
            break
    
    if stop_requested:
        return False
    
    send_to_syslog(f"completed job {job_id}", syslog_port)
    return send_line(tcp_sock, f"DONE {job_id}")


def run_worker(host: str, port: int, output_port: int, syslog_port: int) -> None:
    """Main worker loop."""
    global stop_requested
    
    # Create multicast sender socket using the library
    mcast_sock = multicast.multicastSenderSocket()
    
    while not stop_requested:
        try:
            with socket.create_connection((host, port), timeout=10) as tcp_sock:
                tcp_sock.settimeout(None)
                send_to_syslog("worker started", syslog_port)
                
                while not stop_requested:
                    send_to_syslog("fetching job", syslog_port)
                    
                    if not send_line(tcp_sock, "FETCH"):
                        break
                    response = recv_line(tcp_sock)
                    if response is None:
                        break
                    if response.startswith("JOB "):
                        parts = response.split(maxsplit=2)
                        if len(parts) < 3:
                            continue
                        try:
                            job_id = int(parts[1])
                        except ValueError:
                            continue
                        text = parts[2]
                        if not process_job(mcast_sock, tcp_sock, job_id, text, 
                                         output_port, syslog_port):
                            break
                    elif response == "NOJOB":
                        time.sleep(FETCH_INTERVAL)
                    else:
                        time.sleep(FETCH_INTERVAL)
        except (ConnectionError, OSError):
            send_to_syslog("connection lost, reconnecting", syslog_port)
        if not stop_requested:
            time.sleep(1)
    
    send_to_syslog("worker stopped", syslog_port)
    mcast_sock.close()


def main(argv: list[str]) -> int:
    parser = argparse.ArgumentParser(description="Worker that pulls jobs from the queue")
    parser.add_argument("workqueue", help="work queue host:port (e.g., localhost:50001)")
    parser.add_argument("output_port", type=int, help="UDP multicast port for job output")
    parser.add_argument("syslog_port", type=int, help="UDP port for syslog messages")
    args = parser.parse_args(argv)

    # Parse host:port
    if ":" not in args.workqueue:
        parser.error("workqueue must be in format host:port")
    
    host, port_str = args.workqueue.rsplit(":", 1)
    try:
        port = int(port_str)
    except ValueError:
        parser.error("invalid port in workqueue argument")

    if port <= 0 or port > 65535:
        parser.error("workqueue port must be between 1 and 65535")
    if args.output_port <= 0 or args.output_port > 65535:
        parser.error("output_port must be between 1 and 65535")
    if args.syslog_port <= 0 or args.syslog_port > 65535:
        parser.error("syslog_port must be between 1 and 65535")

    signal.signal(signal.SIGINT, _signal_handler)
    signal.signal(signal.SIGTERM, _signal_handler)

    run_worker(host, port, args.output_port, args.syslog_port)
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))